File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.java
import java.util.*;
class Solution {
    public int[] solution(String today, String[] terms, String[] privacies) {
        // int[] answer = {};
        HashMap<String,Integer> hm = new HashMap();
        ArrayList<Integer> al = new ArrayList<>();
        for(String term : terms){
            StringTokenizer st = new StringTokenizer(term, " ");
            String t = st.nextToken();
            int m = Integer.parseInt(st.nextToken());
            hm.put(t,m);
        }
        int index=1;
        for(String privacy : privacies){
            StringTokenizer st = new StringTokenizer(privacy, " ");
            String date = st.nextToken();
            String term = st.nextToken();
            st = new StringTokenizer(date, ".");
            int y = Integer.parseInt(st.nextToken());
            int m = Integer.parseInt(st.nextToken());
            int d = Integer.parseInt(st.nextToken());
            m+=hm.get(term);
            while(m > 12){
                y++;
                m-=12;
            }
            st = new StringTokenizer(today, ".");
            int ty = Integer.parseInt(st.nextToken());
            int tm = Integer.parseInt(st.nextToken());
            int td = Integer.parseInt(st.nextToken());
            if(ty>y)    al.add(index);
            else if(ty==y && tm > m)    al.add(index);
            else if(ty==y && tm == m && td >= d)    al.add(index);
            index++;
            // System.out.println(privacy);
            // System.out.println(hm);
        }
        int size=0;
        int[] answer = new int[al.size()];
        for(int i : al) answer[size++] = i;
        return answer;
    }
}

File: 프로그래머스/1/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.cpp
#include <string>
#include <vector>
#include <iostream>
#include <unordered_map>
using namespace std;

vector<string> split(string str, char delimeter){
    vector<string> result;
    int start = 0, end;
    while((end = str.find(delimeter, start)) != string::npos){
        result.push_back(str.substr(start, end-start));
        start = end + 1;
    }
    result.push_back(str.substr(start));
    return result;
}

vector<int> solution(string today, vector<string> terms, vector<string> privacies) {
    vector<int> answer;
    unordered_map<string, int> um;
    for(auto s : terms){
        vector<string> v = split(s, ' ');
        um[v[0]] = stoi(v[1]);
    }
    int index = 0;
    for(auto s : privacies){
        index++;
        vector<string> v = split(s, ' ');
        vector<string> date = split(v[0], '.');
        vector<string> date2 = split(today, '.');
        int term = um[v[1]];
        vector<int> dateI;
        vector<int> todayI;
        for(int i=0; i<3; i++){
            dateI.push_back(stoi(date[i]));
            todayI.push_back(stoi(date2[i]));
        }
        dateI[1] += term;
        dateI[0] += ((dateI[1]-1) / 12);
        dateI[1]--;
        dateI[1] %= 12;
        dateI[1]++;
        for(int i=0; i<3; i++){
            if(todayI[i] > dateI[i]){
                answer.push_back(index);
                break;
            }else if(todayI[i] < dateI[i]) break;
            else if(i == 2 && todayI[i] == dateI[i]){
                answer.push_back(index);
                break;
            }
        }
    }
    return answer;
}